design/crd-status.md

```diff
@@ -0,0 +1,316 @@
+    }
+```
+
+**Note:** given the fact that the configuration is parsed both by the speakers and
+by the controller, we might want to expose the status of each component. However,
+there is currently no logic in the configuration parsing that depends on the component.
+For this reason, we can have an initial status that is produced only by the controller
+(which runs in single instance) to validate the status.
+
+As an alternative, we might consider having a state per component, named after the
+single component that produces the status, but in general having a single place to
+check seems more straightforward.
+
+If we go with the _per component_ scenario, we might add a `loadedConfiguration`
+field that exposes the latest loaded configuration by that component. This can't
+be done if we let the controller to expose the single configuration because
+what's loaded might depend on the order the CRs are received with.
+
+#### BGP / BFD Session state
+
+Because of the scalability concerns expressed above, the idea is to produce an instance
+of the resource _per peer / node_, which exposes the state of the session between
+the speaker running on a given node and a given peer.
+
+The name of a given instance will be like `nodename-peer`, and each instance will
+be labeled with the name of the node and the peer it refers to, to make it easier
+to list the status of all the sessions related to a given `BGPPeer` and a given
+node.
+
+```go
+    type MetalLBBGPStatus struct {
+        bgpStatus string
+        bfdStatus string
+    }
+
+    type BGPSessionStatus struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        Status            MetalLBBGPStatus `json:"status,omitempty"`
+    }
+```
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPSessionStatus
+metadata:
+  name: worker0-peer1
+  namespace: metallb-system
+  labels:
+    metallb.io/node: worker0
+    metallb.io/peer: peer1
+status:
+    bgpStatus: Established
+    bfdStatus: Up
+```
+
+The string exposed is taken directly from the output of FRR.
+
+If BFD is not configured for a given `BGPPeer`, the exposed bfdStatus will be "N/A".
+
+**A note about the implementation:** without entering to much into details,
+we will need to implement some sort of polling of the FRR status. Given the
+fact that this CR has no relation with the existing ones, a valid approach is
+to follow what was done for the [metrics exporter](https://github.com/metallb/metallb/blob/main/frr-metrics/exporter.go)
+and have a different component (or even the exporter itself) polling FRR and filling
+the session status. The polling interval must be configurable and large enough to
+avoid impacts both on FRR and on the API.
+The speaker should continue not to have direct interactions with FRR.
+
+#### Service Announcement - BGP
+
+Given a service and a node, we want to expose the `BGPpeer`s the service is configured
```

Comment:
I would prefer to have status in the bgp advertisement crd. For instance, you can list the nodes and the peers that the bgp adv is linked to, as well as, the services. In my opinion, this is the most important crd and what users expect to see.

Note that we can modify the controller to update the status. Not need to do it in each speaker.

Reply:
What users will care about is: why my service is not working?
And, the configuration associated to a given service might come from the union of different bgp advertisements (i.e. multiple peers / multiple nodes / etc).
That's the reason why I think the service must be the key for this status.

Adding the status for all the services associated to a bgp advertisement will give a partial picture and will still make the troubleshooting hard to diagnose.

Reply:
but also here is incomplete. It is possible to see the nodes and peers but the user won't know where bgp advs are causing that configuration

Reply:
Then we could add the list of advs that contribute to a given config. But this version is still preferrable imo

Reply:
what about having the 2 of them?

Reply:
let's stick with the current proposal as it can be a starting point that we extend.
We can expand, but when we push an api it's really hard to get back.